diff --git a/packages/server/src/features/languageFeatures.ts b/packages/server/src/features/languageFeatures.ts
index 6f347971..d0bcdf97 100644
--- a/packages/server/src/features/languageFeatures.ts
+++ b/packages/server/src/features/languageFeatures.ts
@@ -43,6 +43,9 @@ export function register(
 				};
 			},
 		);
+		if (list) list.items.forEach((item, i) => {
+			item.sortText = `${item.sortText ?? ''}${i}`
+		})
 		const insertReplaceSupport = params.capabilities.textDocument?.completion?.completionItem?.insertReplaceSupport ?? false;
 		if (!insertReplaceSupport && list) {
 			for (const item of list.items) {
