diff --git a/packages/vue-typescript/src/use/useSfcTemplateScript.ts b/packages/vue-typescript/src/use/useSfcTemplateScript.ts
index 48d02685..7f50d8e6 100644
--- a/packages/vue-typescript/src/use/useSfcTemplateScript.ts
+++ b/packages/vue-typescript/src/use/useSfcTemplateScript.ts
@@ -49,10 +49,8 @@ export function useSfcTemplateScript(
 	const cssScopedClasses = computed(() => {
 		const obj: Record<string, ReturnType<typeof cssClasses.findClassNames>> = {};
 		for (const style of styleDocuments.value) {
-			if (style.scoped) {
 				const classes = getCssClasses(style.textDocument);
 				obj[style.textDocument.uri] = classes;
-			}
 		}
 		return obj;
 	});
