diff --git a/packages/vue-typescript/src/use/useSfcTemplateScript.ts b/packages/vue-typescript/src/use/useSfcTemplateScript.ts
index 8fd58c33..c1b76e3d 100644
--- a/packages/vue-typescript/src/use/useSfcTemplateScript.ts
+++ b/packages/vue-typescript/src/use/useSfcTemplateScript.ts
@@ -46,7 +46,7 @@ export function useSfcTemplateScript(
 	const cssScopedClasses = computed(() => {
 		const obj: Record<string, Record<string, TextRange[]>> = {};
 		for (const style of styleFiles.value) {
-			if (style.data.scoped || compilerOptions.experimentalResolveNonScopedCssClasses) {
+			if (style.data.scoped || true) {
 				const classes = getCssClasses(style);
 				obj[style.fileName] = classes;
 			}
